src/vi/goctruyentranhvui/src/eu/kanade/tachiyomi/extension/vi/goctruyentranhvui/GocTruyenTranhVui.kt

```diff
@@ -0,0 +1,252 @@
+import eu.kanade.tachiyomi.source.model.SChapter
+import eu.kanade.tachiyomi.source.model.SManga
+import eu.kanade.tachiyomi.source.online.HttpSource
+import eu.kanade.tachiyomi.util.asJsoup
+import keiyoushi.utils.parseAs
+import okhttp3.Headers
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import java.text.SimpleDateFormat
+import java.util.Calendar
+import java.util.Locale
+
+class GocTruyenTranhVui() : HttpSource() {
+    override val lang = "vi"
+
+    override val baseUrl = "https://goctruyentranhvui17.com"
+
+    override val name = "Goc Truyen Tranh Vui"
+
+    private val apiUrl = "$baseUrl/api/v2"
+
+    private val apiComic = "$baseUrl/api/comic"
+
+    private val dateFormat = SimpleDateFormat("dd-MM-yyyy", Locale.US)
+
+    override val supportsLatest: Boolean = true
+
+    override val client: OkHttpClient = network.cloudflareClient.newBuilder()
+        .rateLimit(3)
+        .build()
+
+    override fun headersBuilder(): Headers.Builder = super.headersBuilder()
+        .add("Referer", "$baseUrl/")
+
+    override fun chapterListParse(response: Response): List<SChapter> {
+        val res = response.asJsoup()
+        val mangaIdRes = res.selectFirst("input[id=comic-id-comment]")!!.attr("value") // id
+        val slug = response.request.url.toString().substringAfterLast("/") // slug
+        val mangaId = MangaIdCache.map[slug] ?: mangaIdRes // id
+
+        val urlChapter = apiComic.toHttpUrl().newBuilder().apply {
+            addPathSegments(mangaId)
+            addPathSegments("chapter")
+            addQueryParameter("limit", "-1")
+        }.build()
+        val getChapter = client.newCall(GET(urlChapter, headers))
+        val chapter = getChapter.execute().parseAs<ListChapter>()
+        return chapter.result.chapters.map { it.toChapter(slug) }.ifEmpty {
+            // Get chapter from manga page if chapterJson doesn't have chapter
+            res.select(".chapter-list .list .col-md-6").map { itm ->
+                SChapter.create().apply {
+                    name = itm.select("a .chapter-info").text()
+                    date_upload = parseDate(itm.select(".col-md-6 .text--disabled .d-flex").text())
+                    itm.selectFirst("a")?.absUrl("href")?.let { setUrlWithoutDomain(it) }
+                }
+            }
+        }
+    }
+
+    private fun parseDate(date: String): Long = runCatching {
+        val calendar = Calendar.getInstance()
+        val number = date.replace(Regex("[^0-9]"), "").trim().toInt()
+        when (date.replace(Regex("[0-9]"), "").trim()) {
+            "phút trước" -> calendar.apply { add(Calendar.MINUTE, -number) }.timeInMillis
+            "giờ trước" -> calendar.apply { add(Calendar.HOUR, -number) }.timeInMillis
+            "ngày trước" -> calendar.apply { add(Calendar.DAY_OF_YEAR, -number) }.timeInMillis
+            else -> dateFormat.parse(date)?.time
```

Comment:
```suggestion
            else -> dateFormat.tryParse(date)
```
need to import, and remove the runcatching